iree/compiler/JitEval/Runtime.cpp

```diff
@@ -0,0 +1,232 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/JitEval/Runtime.h"
+
+#include "iree/compiler/Dialect/VM/Target/Bytecode/BytecodeModuleTarget.h"
+#include "iree/hal/buffer_view.h"
+#include "iree/hal/vmvx/registration/driver_module.h"
+#include "iree/modules/hal/module.h"
+#include "iree/vm/ref_cc.h"
+#include "mlir/IR/BuiltinOps.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace JitEval {
+
+namespace {
+
+Type mapElementType(Location loc, iree_hal_element_type_t halElementType) {
+  Builder builder(loc.getContext());
+  if (iree_hal_element_numerical_type_is_integer(halElementType)) {
+    return builder.getIntegerType(iree_hal_element_bit_count(halElementType));
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_FLOAT_32) {
+    return builder.getF32Type();
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_FLOAT_64) {
+    return builder.getF64Type();
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_FLOAT_16) {
+    return builder.getF16Type();
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_BFLOAT_16) {
+    return builder.getBF16Type();
+  }
+
+  emitError(loc) << "unrecognized evaluated buffer view element type: "
+                 << halElementType;
+  return {};
+}
+
+}  // namespace
+
+CompiledBinary::CompiledBinary() {
+  IREE_CHECK_OK(iree_vm_instance_create(iree_allocator_system(), &instance));
+}
+
+CompiledBinary::~CompiledBinary() {
+  iree_vm_context_release(context);
+  iree_vm_module_release(main_module);
+  iree_vm_instance_release(instance);
+}
+
+LogicalResult CompiledBinary::invokeNullary(Location loc, StringRef name,
+                                            ResultsCallback callback) {
+  iree_vm_function_t function;
+  IREE_CHECK_OK(main_module->lookup_function(
```

Comment:
it'd be good to handle these gracefully - otherwise an invalid user program that we extract would crash the compiler
you could make these functions return Status (from iree/base/status_cc.h) and then have the top level thing `return LogicalResult(status.ok();` (or whatever that bool-accepting variant is)

Reply:
Oh yeah - I'll be going through and proper error handling everything that is realistically failable. Just need to write some status bridgy thingy.